src/main/java/org/breedinginsight/services/OntologyService.java

```diff
@@ -234,4 +241,86 @@ public void revokeOntology(@NotNull UUID programId, @NotNull UUID sharedProgramI
         // Remove record from db
         programOntologyDAO.revokeSharedOntology(sharedOntology);
     }
+
+    public SubscribedProgram subscribeOntology(UUID programId, UUID sharingProgramId) throws DoesNotExistException, UnprocessableEntityException {
+        // Check that program exists
+        Program program = getProgram(programId);
+
+        // Check that shared program exists
+        Optional<ProgramSharedOntologyEntity> optionalSharedOntology = programOntologyDAO.getSharedOntologyById(sharingProgramId, programId);
+        if (optionalSharedOntology.isEmpty()) {
+            throw new DoesNotExistException("Shared ontology between specified programs was not found.");
+        }
+        ProgramSharedOntologyEntity sharedOntology = optionalSharedOntology.get();
+
+        // Check that program does not have any traits of its own
+        List<Trait> traits = traitDAO.getTraitsByProgramId(programId);
+        if (traits.size() > 0) {
+            throw new UnprocessableEntityException("Program already has traits, cannot subscribe to a shared ontology");
+        }
+
+        // Check that program does not have any current shares of its own ontology
+        List<ProgramSharedOntologyEntity> sharedOntologies = programOntologyDAO.getSharedOntologies(program.getId());
+        if (!sharedOntologies.isEmpty()) {
+            throw new UnprocessableEntityException("Program has shared its ontology with other programs, cannot subscribe to another ontology.");
+        }
+
+
+        // Subscribe
+        programOntologyDAO.acceptSharedOntology(sharedOntology);
+
+        // Get the subscription record
+        try {
+            List<SubscribedProgram> subscribedProgramOptions = getSubscribeOntologyOptions(programId);
+            for (SubscribedProgram option: subscribedProgramOptions) {
+                if (option.getProgramId().equals(sharingProgramId)) {
+                    return option;
+                }
+            }
+        } catch (DoesNotExistException e) {
+            throw new InternalServerException("Recently subscribed program cannot be found.");
+        }
+
+        throw new InternalServerException("Recently subscribed program cannot be found.");
+    }
+
+    public void unsubscribeOntology(UUID programId, UUID sharingProgramId) throws DoesNotExistException, UnprocessableEntityException {
+        // Check that program exists
+        Program program = getProgram(programId);
+
+        // Check that shared program exists
+        Optional<ProgramSharedOntologyEntity> optionalSharedOntology = programOntologyDAO.getSharedOntologyById(sharingProgramId, programId);
+        if (optionalSharedOntology.isEmpty()) {
+            throw new DoesNotExistException("Shared ontology between specified programs was not found.");
```

Comment:
Same as above...thoughts on this throwing some sort of illegal state type of exception?